runtime/js-compute-runtime/builtins/crypto-algorithm.cpp

```diff
@@ -779,6 +935,291 @@ CryptoAlgorithmImportKey::normalize(JSContext *cx, JS::HandleValue value) {
+                                                             CryptoKeyUsages usages) {
+  MOZ_ASSERT(cx);
+  JS::RootedObject result(cx);
+
+  // 2. If usages contains an entry which is not "sign" or "verify", then throw a SyntaxError.
+  if (!usages.canOnlySignOrVerify()) {
+    // TODO Rename error to SyntaxError
+    JS_ReportErrorLatin1(cx, "HMAC keys only support 'sign' and 'verify' operations");
+    return nullptr;
+  }
+
+  std::unique_ptr<std::span<uint8_t>> data;
+  // 3. Let hash be a new KeyAlgorithm.
+  // 4.
+  switch (format) {
+  // 5. If format is "raw":
+  case CryptoKeyFormat::Raw: {
+    // 5.1 Let data be the octet string contained in keyData.
+    data = std::make_unique<std::span<uint8_t>>(std::get<std::span<uint8_t>>(keyData));
+    if (!data) {
+      JS_ReportErrorLatin1(cx, "Supplied keyData must be either an ArrayBuffer, TypedArray, or DataView.");
+      convertErrorToDataError(cx);
+      return nullptr;
+    }
+
+    // 5.2 Set hash to equal the hash member of normalizedAlgorithm.
+    // Step 5.2 is done in the call to CryptoKey::createHMAC
+    break;
+  }
+  //6.  If format is "jwk":
+  case CryptoKeyFormat::Jwk: {
+    // 6.1 If keyData is a JsonWebKey dictionary:
+      // 6.2 Let jwk equal keyData.
+    // Otherwise:
+      // 6.3 Throw a DataError.
+    auto jwk = std::get<JsonWebKey *>(keyData);
+    if (!jwk) {
+      JS_ReportErrorLatin1(cx, "Supplied format is not a JSONWebKey");
+      return nullptr;
+    }
+    // 6.4 If the kty field of jwk is not "oct", then throw a DataError.
+    // Step 6.4 has already been done in the other implementation of
+    // CryptoAlgorithmHMAC_Import::importKey which is called before this one.
+    // 6.5 If jwk does not meet the requirements of Section 6.4 of JSON Web Algorithms [JWA], then throw a DataError.
+
+    // 6.6 Let data be the octet string obtained by decoding the k field of jwk.
+    auto dataResult = GlobalProperties::forgivingBase64Decode(
+      jwk->k.value(), GlobalProperties::base64URLDecodeTable);
+    if (dataResult.isErr()) {
+      JS_ReportErrorLatin1(cx,
+                         "The JWK member 'k' could not be base64url decoded or contained padding");
+      convertErrorToDataError(cx);
+      return nullptr;
+    }
+    auto data_string = dataResult.unwrap();
+    auto ddd = std::vector<uint8_t>(data_string.begin(), data_string.end());
+    data = std::make_unique<std::span<uint8_t>>(
+      ddd
+    );
+    // 6.7 Set the hash to equal the hash member of normalizedAlgorithm.
+    CryptoAlgorithmIdentifier hash = this->hashIdentifier;
+    switch(hash) {
+      // 6.8 If the name attribute of hash is "SHA-1":
+      case CryptoAlgorithmIdentifier::SHA_1: {
+        // If the alg field of jwk is present and is not "HS1", then throw a DataError.
+        if (jwk->alg.has_value() && jwk->alg.value() != "HS1") {
+          JS_ReportErrorLatin1(cx, "1Operation not permitted");
```

Comment:
Are these number prefixes intended to stick around?

Reply:
```suggestion
          JS_ReportErrorLatin1(cx, "Operation not permitted");
```